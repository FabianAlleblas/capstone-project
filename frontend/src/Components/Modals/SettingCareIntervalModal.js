import PropTypes from 'prop-types'
import styled from 'styled-components'
import useForm from '../../hooks/useForm'
import Button from '../Buttons/Button'
import { CloseIcon } from '../Icons'

SettingCareIntervalModal.propTypes = {
  isFertilizer: PropTypes.bool,
  setShowCareSetting: PropTypes.func.isRequired,
  updateCareInterval: PropTypes.func.isRequired,
  plantId: PropTypes.number.isRequired,
  className: PropTypes.string,
}

export default function SettingCareIntervalModal({
  isFertilizer,
  setShowCareSetting,
  updateCareInterval,
  plantId,
  className,
}) {
  const { handleInputChange, formData } = useForm()

  return (
    <Wrapper className={className}>
      <Modal>
        <ClosingButton onClick={closeCareSetting}>
          <CloseIcon />
        </ClosingButton>
        <Form onSubmit={saveCareInterval}>
          <Label>
            Set your preferred {isFertilizer ? 'fertilizing' : 'watering'}{' '}
            interval (default: {isFertilizer ? '4 weeks' : '10 days'}):
            <Input
              name={isFertilizer ? 'fertilizerInterval' : 'waterInterval'}
              type="number"
              placeholder="1-99"
              min="1"
              max="99"
              onChange={handleInputChange}
              required
            />
          </Label>
          <ButtonStyled>Save</ButtonStyled>
        </Form>
      </Modal>
    </Wrapper>
  )

  function saveCareInterval(event) {
    event.preventDefault()
    updateCareInterval(plantId, formData)
    setShowCareSetting()
  }

  function closeCareSetting() {
    setShowCareSetting()
  }
}

const Wrapper = styled.div`
  background-color: rgba(0, 0, 0, 0.4);
  display: grid;
  place-items: center;
  width: 100%;
  z-index: 10;
`

const Modal = styled.div`
  background-color: var(--primary-dark);
  border-radius: 20px;
  box-shadow: 4px 4px 10px var(--shadow);
  padding: 10px;
  width: 80%;
`

const ClosingButton = styled.button`
  border: none;
  background-color: transparent;
  width: 18px;
`

const Form = styled.form`
  display: grid;
  gap: 20px;
  justify-items: center;
  padding: 14px 10px 10px;
`
const Label = styled.label`
  color: var(--form-font-color);
  display: flex;
  flex-wrap: wrap;
  font-size: 1.2rem;
  justify-content: center;
  text-align: center;
`

const Input = styled.input`
  border-radius: 18px;
  border: none;
  color: var(--form-font-color);
  font-family: 'Josefin sans', sans-serif;
  font-size: 1rem;
  margin-top: 20px;
  padding: 8px 14px;
  width: 40%;

  &:focus {
    box-shadow: 0 0 4px var(--shadow) inset;
    outline: none;
  }
`

const ButtonStyled = styled(Button)`
  scale: 0.8;
  width: 75%;
`
