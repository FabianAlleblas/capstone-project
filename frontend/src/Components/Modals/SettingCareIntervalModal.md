Modal: Water Interval

```js
<SettingCareIntervalModal onSubmit={() => {}} onClick={() => {}} />
```

Modal: Fertilizer Interval

```js
<SettingCareIntervalModal isFertilizer onSubmit={() => {}} onClick={() => {}} />
```
