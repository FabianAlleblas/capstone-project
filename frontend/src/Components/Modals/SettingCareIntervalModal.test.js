import { render } from '@testing-library/react'
import user from '@testing-library/user-event'
import 'jest-styled-components'
import SettingCareIntervalModal from './SettingCareIntervalModal'

const setShowCareSettingMock = jest.fn()
const updateCareIntervalMock = jest.fn()

describe('SettingCareIntervalModal', () => {
  it('renders correctly', () => {
    const { container } = render(
      <SettingCareIntervalModal
        setShowCareSetting={setShowCareSettingMock}
        updateCareInterval={updateCareIntervalMock}
        plantId={1}
      />
    )

    expect(container.firstChild).toMatchSnapshot()
  })
  it('calls updateCareIntervalMock with the correct data and calls also setShowCareSettingMock by clicking the save button', () => {
    const { getByPlaceholderText, getByText } = render(
      <SettingCareIntervalModal
        setShowCareSetting={setShowCareSettingMock}
        updateCareInterval={updateCareIntervalMock}
        plantId={1}
      />
    )

    user.type(getByPlaceholderText('1-99'), '10')
    user.click(getByText('Save'))

    expect(updateCareIntervalMock).toHaveBeenCalledWith(1, {
      waterInterval: '10',
    })
    expect(setShowCareSettingMock).toHaveBeenCalled()
  })
  it('calls setShowCareSettingMock by clicking the close button', () => {
    const { getByRole } = render(
      <SettingCareIntervalModal
        setShowCareSetting={setShowCareSettingMock}
        updateCareInterval={updateCareIntervalMock}
        plantId={1}
      />
    )

    user.click(getByRole('button', { name: /CloseIcon/i }))

    expect(setShowCareSettingMock).toHaveBeenCalled()
  })
  it('renders with the correct text and calls updateCareIntervalMock with the correct data', () => {
    const { getByPlaceholderText, getByLabelText, getByText } = render(
      <SettingCareIntervalModal
        setShowCareSetting={setShowCareSettingMock}
        updateCareInterval={updateCareIntervalMock}
        plantId={1}
        isFertilizer
      />
    )

    user.type(getByPlaceholderText('1-99'), '10')
    user.click(getByText('Save'))

    expect(updateCareIntervalMock).toHaveBeenCalledWith(1, {
      fertilizerInterval: '10',
    })

    expect(
      getByLabelText(/Set your preferred fertilizing interval/i)
    ).toBeInTheDocument()
  })
})
