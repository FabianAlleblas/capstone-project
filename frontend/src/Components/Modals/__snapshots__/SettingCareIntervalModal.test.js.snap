// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingCareIntervalModal renders correctly 1`] = `
.c6 {
  background-color: var(--primary-light);
  border-radius: 22px;
  border: none;
  box-shadow: 0 0 4px var(--shadow);
  color: var(--primary-dark);
  font-family: 'Josefin sans';
  font-size: 1.25rem;
  font-weight: 700;
  padding: 12px 16px 8px;
  width: 100%;
}

.c0 {
  background-color: rgba(0,0,0,0.4);
  display: grid;
  place-items: center;
  width: 100%;
  z-index: 10;
}

.c1 {
  background-color: var(--primary-dark);
  border-radius: 20px;
  box-shadow: 4px 4px 10px var(--shadow);
  padding: 10px;
  width: 80%;
}

.c2 {
  border: none;
  background-color: transparent;
  width: 24px;
}

.c3 {
  display: grid;
  gap: 20px;
  justify-items: center;
  padding: 14px 10px 10px;
}

.c4 {
  color: var(--form-font-color);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  font-size: 1.2rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  text-align: center;
}

.c5 {
  border-radius: 18px;
  border: none;
  color: var(--form-font-color);
  font-family: 'Josefin sans',sans-serif;
  font-size: 1rem;
  margin-top: 20px;
  padding: 8px 14px;
  width: 40%;
}

.c5:focus {
  box-shadow: 0 0 4px var(--shadow) inset;
  outline: none;
}

.c7 {
  scale: 0.8;
  width: 75%;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <button
      class="c2"
    >
      <svg>
        closeicon.svg
      </svg>
    </button>
    <form
      class="c3"
    >
      <label
        class="c4"
      >
        Set your preferred 
        watering
         
        interval (default: 
        10 days
        ):
        <input
          class="c5"
          max="99"
          min="1"
          name="waterInterval"
          placeholder="1-99"
          required=""
          type="number"
        />
      </label>
      <button
        class="c6 c7"
      >
        Save
      </button>
    </form>
  </div>
</div>
`;
